GoogleDataTransport/GDTCORLibrary/GDTCORDatabase.m

```diff
@@ -0,0 +1,316 @@
+    return NO;
+  }
+  if (stmtCache == nil) {
+    GDTCORLogError(GDTCORMCEDatabaseError, @"%@", @"Statement cache needs to be non-nil");
+    return NO;
+  }
+  sqlite3_stmt *stmt =
+      (sqlite3_stmt *)CFDictionaryGetValue(stmtCache, (__bridge const void *)(sql));
+  if (stmt == NULL) {
+    if (GDTCORSQLCompileSQL(&stmt, db, sql)) {
+      if (stmtCache != nil && cacheStmt) {
+        CFDictionaryAddValue(stmtCache, (__bridge const void *)(sql), stmt);
+      }
+    } else {
+      GDTCORLogError(GDTCORMCEDatabaseError, @"SQL did not compile: %@", sql);
+      return NO;
+    }
+  }
+  [bindings enumerateKeysAndObjectsUsingBlock:^(NSNumber *_Nonnull key, NSString *_Nonnull obj,
+                                                BOOL *_Nonnull stop) {
+    GDTCORSQLBindObjectToParam(stmt, key.intValue, obj);
+  }];
+  BOOL result = GDTCORSQLRunQuery(db, stmt, eachRow);
+  if (cacheStmt) {
+    GDTCORSQLReset(stmt);
+  } else {
+    GDTCORSQLFinalize(stmt);
+  }
+  return result;
+}
+
+/** Runs a series of migrations, setting the user_version PRAGMA to the highest version number.
+ *
+ * @param db The db to operate on.
+ * @param migrations The map of user_version numbers to the SQL statement needed to update to that
+ *   version.
+ * @param stmtCache The statement cache.
+ */
+static void RunMigrations(sqlite3 *db,
+                          NSDictionary<NSNumber *, NSString *> *migrations,
+                          CFMutableDictionaryRef stmtCache) {
+  if (!db) {
+    GDTCORLogError(GDTCORMCEDatabaseError, @"%@", @"The database is closed");
+    return;
+  }
+  int userVersion = GetUserVersion(db);
+  __block int newUserVersion = userVersion;
+  [migrations enumerateKeysAndObjectsUsingBlock:^(NSNumber *_Nonnull key, NSString *_Nonnull obj,
+                                                  BOOL *_Nonnull stop) {
+    if (key.intValue > newUserVersion) {
+      if (RunNonQuery(db, nil, obj, stmtCache, NO)) {
+        newUserVersion = key.intValue;
+      } else {
+        GDTCORLogError(GDTCORMCEDatabaseError, @"%@", @"Migration failed: version:%@ statement:%@",
+                       key, obj);
+      }
+    }
+  }];
+
+  if (newUserVersion != userVersion) {
+    SetUserVersion(db, newUserVersion);
+  }
+}
+
+#pragma mark - GDTCORDatabase
+
+@implementation GDTCORDatabase {
+  /** The sqlite database. */
+  sqlite3 *_db;
+}
+
+- (nullable instancetype)initWithURL:(nullable NSURL *)dbFileURL
+                 migrationStatements:
+                     (nonnull NSDictionary<NSNumber *, NSString *> *)migrationStatements {
+  // Only 1 instance per file is allowed. Return nil if one is already open.
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    instancesToDBFiles = [[NSMutableDictionary alloc] init];
+  });
+  __block GDTCORDatabase *extantDB;
+  dispatch_barrier_sync(dispatch_get_global_queue(QOS_CLASS_DEFAULT, 0), ^{
+    extantDB = instancesToDBFiles[dbFileURL.path];
+  });
+  if (extantDB) {
+    return nil;
```

Comment:
No, I'd rather it return nil so that you know there's already a db instance working on that file. I don't want the same db instance being operated from two different places.